src/host_beta/messageRouter.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import {
+    encodeMessageForChannel,
+    FrameToolsEvent,
+    IOpenEditorData,
+    parseMessageFromChannel,
+    TelemetryData,
+    WebSocketEvent,
+    WebviewEvent,
+} from '../common/webviewEvents';
+
+declare const acquireVsCodeApi: () => {postMessage(message: unknown, args?: any|undefined): void};
+const vscode = acquireVsCodeApi();
+
+/**
+ * Both the DevTools iframe and the Extension (which owns the ws connection to the browser) will
+ * post messages to the Webview window for communication between the two. This class routes message
+ * to the correct location based on the origin and type of message posted.
+ *
+ */
+export class MessageRouter {
+    private toolsFrameWindow: Window | undefined;
+
+    constructor(webviewWindow: Window | null) {
+        if (!webviewWindow) {
+            return;
+        }
+
+        webviewWindow.addEventListener('DOMContentLoaded', () => {
+            const tw = (document.getElementById('host') as HTMLIFrameElement).contentWindow;
+            this.toolsFrameWindow = tw || undefined;
+        });
+
+
+        const extensionMessageCallback = this.onMessageFromChannel.bind(this);
+
+        // Both the DevTools iframe and the extension will post messages to the webview
+        // Listen for messages and forward to correct recipient based on origin
+        webviewWindow.addEventListener('message', messageEvent => {
+            const fromExtension = messageEvent.origin.startsWith('vscode-webview://');
+            if (!fromExtension) {
+                // Send message from DevTools to Extension
+                this.onMessageFromFrame(messageEvent.data.method as FrameToolsEvent, messageEvent.data.args as any[]);
+            } else if (this.toolsFrameWindow) {
+                // Send message from Extension to DevTools
+                parseMessageFromChannel(
+                    messageEvent.data,
+                    extensionMessageCallback,
+                );
+                messageEvent.preventDefault();
+                messageEvent.stopImmediatePropagation();
+                return false;
```

Comment:
not sure if this return statement is needed

Reply:
Good call. Not sure why this was present in the legacy code, but I don't think return values in event handlers are ever used